### PR TITLE
[DRAFT] Restrict posix passwords on ldap bind with config

### LINKED ISSUE #2027: Prevent posix password from operating with LDAP without a config option
Once we have #41, I think that we should guard posix password checking via LDAP with a config/domain config flag. This is because app passwords are a better solution to legacy application integration, and we don't want people to reach for these in common cases. They should be a "last resort".

### PR DESCRIPTION
**Pull request by NavinShrinivas**
draft fix #2027 
A draft PR, tests fmt clippy have not been confirmed yet.
Checklist

- [x] This pr contains no AI generated code
- [ ] cargo fmt has been run
- [ ] cargo clippy has been run
- [ ] cargo test has been run and passes
- [ ] book chapter included (if relevant)
- [ ] design document included (if relevant)

### CODE REVIEW

File: server/lib/src/constants/entries.rs [hunk 1]
```diff
@@ -361,6 +363,18 @@ impl From<Attribute> for &'static str {
     fn from(val: Attribute) -> Self {
         match val {
             Attribute::Account => ATTR_ACCOUNT,
+            Attribute::SystemMay => ATTR_SYSTEMMAY,
```
Comment (Firstyear) [inline server/lib/src/constants/entries.rs:366]: @yaleman ?

File: server/lib/src/constants/schema.rs [hunk 2]
```diff
@@ -110,6 +110,16 @@ pub static ref SCHEMA_ATTR_DOMAIN_NAME: SchemaAttribute = SchemaAttribute {
     syntax: SyntaxType::Utf8StringIname,
     ..Default::default()
 };
+pub static ref SCHEMA_ATTR_DOMAIN_LDAP_ALLOW_UNIX_PW_BIND: SchemaAttribute = SchemaAttribute {
+    uuid: UUID_SCHEMA_ATTR_DOMAIN_LDAP_ALLOW_UNIX_PW_BIND,
+    name: Attribute::DomainLdapAllowUnixPwBind.into(),
+    description: "Configuration to allow binds to LDAP objects using UNIX passwords.".to_string(),
+
+    index: vec![IndexType::Equality, IndexType::Presence],
```
Comment (Firstyear) [inline server/lib/src/constants/schema.rs:118]: I don't think this needs indexing at all, just unset this.

File: server/lib/src/constants/schema.rs [hunk 3]
```diff
@@ -110,6 +110,16 @@ pub static ref SCHEMA_ATTR_DOMAIN_NAME: SchemaAttribute = SchemaAttribute {
...
+    unique: true,
```
Comment (Firstyear) [inline server/lib/src/constants/schema.rs:119]: can't be unique, unset this.

File: server/lib/src/server/mod.rs [hunk 4]
```diff
@@ -793,6 +794,14 @@ pub trait QueryServerTransaction<'a> {
             })
     }
 
+    fn get_domain_ldap_allow_unix_pw_bind(&mut self) -> Result<bool, OperationError> {
+        let res = self.internal_search_uuid(UUID_DOMAIN_INFO);
+        match res.unwrap().get_ava_single("domain_ldap_allow_unix_pw_bind"){
```
Comment (yaleman) [inline server/lib/src/server/mod.rs:799]: needs to be the const, not a string

File: server/lib/src/idm/ldap.rs [hunk 5]
```diff
@@ -1,4 +1,4 @@
-//! LDAP specific operations handling components. This is where LDAP operations
+//! LDAP specific operations handling components. This is where LDAP operationsldap.rs
```
Comment (yaleman) [inline server/lib/src/idm/ldap.rs:1]: Not sure where this came from? :)

Comment (Firstyear) [inline server/lib/src/idm/ldap.rs:1]: typo?

**Comment by NavinShrinivas** [comment]
Ignore the random prints, the code is very rough.....only till I get the functionality down

File: server/lib/src/idm/ldap.rs (hunk 5 above)
Comment (NavinShrinivas) [reply server/lib/src/idm/ldap.rs:1]: me spamming keybinds on nvim xD

File: server/lib/src/server/mod.rs (hunk 4 above)
Comment (NavinShrinivas) [reply server/lib/src/server/mod.rs:799]: you mean the attribute in `get_ava_single()`? Should I be passing the in const that represents the attrribute?

File: server/lib/src/idm/ldap.rs [hunk 6]
```diff
@@ -445,18 +445,36 @@ impl LdapServer {
...
+            debug!(
+                "here {:?}",
+                idms.proxy_read()
+                    .await
+                    .qs_read
+                    .d_info
+                    .d_ldap_allow_unix_pw_bind
+            );
+            let lae = LdapAuthEvent::from_parts(target_uuid, pw.to_string())?;
+            idm_auth.auth_ldap(&lae, ct).await.and_then(|r| {
```
Comment (Firstyear) [inline server/lib/src/idm/ldap.rs:464]: Why not read this setting for allowing the unique bind here inside of auth_ldap? I think because youi have the proxy read open above that's causing a double-read transaction which is wrong. There should only be *one* transaction per operation. So I'd revert the changes just here, and look into checking this flag in auth_ldap.

File: server/lib/src/idm/ldap.rs [hunk 7]
```diff
@@ -643,15 +661,19 @@ mod tests {
...
         assert!(idms_prox_write.commit().is_ok());
-
         let anon_t = ldaps.do_bind(idms, "", "").await.unwrap().unwrap();
         assert!(anon_t.effective_session == LdapSession::UnixBind(UUID_ANONYMOUS));
         assert!(
             ldaps.do_bind(idms, "", "test").await.unwrap_err() == OperationError::NotAuthenticated
         );
 
         // Now test the admin and various DN's
+        // Writing test for optional use of POSIX password for bind
```
Comment (Firstyear) [inline server/lib/src/idm/ldap.rs:676]: Probably should make a new test case here, that also tests when the condition is *false*.

File: server/lib/src/server/mod.rs [hunk 8]
```diff
@@ -1500,8 +1510,20 @@ impl<'a> QueryServerWriteTransaction<'a> {
     pub(crate) fn reload_domain_info(&mut self) -> Result<(), OperationError> {
         let domain_name = self.get_db_domain_name()?;
         let display_name = self.get_db_domain_display_name()?;
+        let domain_ldap_allow_unix_pw_bind = match self.get_domain_ldap_allow_unix_pw_bind(){
+            Ok(v) => v,
+            _ => {
+                admin_warn!("Defaulting ldap_allow_unix_pw_bind to false");
+                self.set_domain_ldap_allow_unix_pw_bind(false)?;
+                false
```
Comment (Firstyear) [inline server/lib/src/server/mod.rs:1518]: For now we have to default it to true here. This is because if we defaulted it to false right now, we'd break ldap for a lot of people.

File: server/lib/src/server/mod.rs [hunk 9]
```diff
@@ -1500,8 +1510,20 @@ impl<'a> QueryServerWriteTransaction<'a> {
...
+            }
+        };
         let domain_uuid = self.be_txn.get_db_d_uuid();
         let mut_d_info = self.d_info.get_mut();
+        if mut_d_info.d_ldap_allow_unix_pw_bind != domain_ldap_allow_unix_pw_bind{
+            admin_warn!("Using of unix pw for ldap bind is now allowed : {}", domain_ldap_allow_unix_pw_bind);
```
Comment (Firstyear) [inline server/lib/src/server/mod.rs:1524]: I don't understand this message and why it was added?

File: server/lib/src/server/mod.rs (hunk 4 above)
Comment (yaleman) [reply server/lib/src/server/mod.rs:799]: You probably need to rebase, `get_ava_single` should only take an `Attribute`

File: server/lib/src/server/mod.rs [hunk 10]
```diff
@@ -1537,7 +1559,15 @@ impl<'a> QueryServerWriteTransaction<'a> {
         let filt = filter_all!(f_eq(Attribute::Uuid, udi));
         self.internal_modify(&filt, &modl)
     }
-
+    pub fn set_domain_ldap_allow_unix_pw_bind(
```
Comment (Firstyear) [inline server/lib/src/server/mod.rs:1562]: I don't think you need a dedicated set function here, none of the other domain update bits have it.

**Review by Firstyear — CHANGES_REQUESTED** [review]

File: server/lib/src/constants/entries.rs (hunk 1 above)
Comment (yaleman) [reply server/lib/src/constants/entries.rs:366]: Someone needs to rebase on `HEAD`?

File: server/lib/src/server/mod.rs (hunk 4 above)
Comment (NavinShrinivas) [reply server/lib/src/server/mod.rs:799]: Makes sense, on rebasing, get_ava_single accpets Attributes

File: server/lib/src/idm/ldap.rs [hunk 11]
```diff
@@ -642,16 +645,35 @@ mod tests {
 
         let pce = UnixPasswordChangeEvent::new_internal(UUID_ADMIN, TEST_PASSWORD);
 
-        assert!(idms_prox_write.set_unix_account_password(&pce).is_ok());
+        let allow_unix_pw_flag = ModifyEvent::new_internal_invalid(
+            filter!(f_eq(Attribute::Uuid, PartialValue::Uuid(UUID_DOMAIN_INFO))),
+            ModifyList::new_purge_and_set(Attribute::DomainLdapAllowUnixPwBind, Value::Bool(false)),
+        );
+        assert!(idms_prox_write.qs_write.modify(&allow_unix_pw_flag).is_ok());
```
Comment (NavinShrinivas) [inline server/lib/src/idm/ldap.rs:652]: setting the flag to false here doesnt seem to block the bind in auth ldap...i.e the test on line 666 id failign

File: server/lib/src/idm/server.rs [hunk 12]
```diff
@@ -1361,6 +1361,10 @@ impl<'a> IdmServerAuthTransaction<'a> {
             admin_error!("Failed to start auth ldap -> {:?}", e);
             e
         })?;
+        if self.qs_read.d_info.d_ldap_allow_unix_pw_bind == false {
```
Comment (NavinShrinivas) [inline server/lib/src/idm/server.rs:1364]: this is where the condition should be read as false, but currently is not

File: server/lib/src/idm/ldap.rs [hunk 13]
```diff
@@ -642,16 +645,35 @@ mod tests {
...
             ldaps.do_bind(idms, "", "test").await.unwrap_err() == OperationError::NotAuthenticated
         );
 
         // Now test the admin and various DN's
+
+
+        // Writing test for optional use of POSIX password for bind
+        // NOT ALLOWING UNIX_PW FOR BIND :
+        let admin_t = ldaps.do_bind(idms, "admin", TEST_PASSWORD).await.unwrap(); // Cannot unwrap futher as on disallowed unix_pw_bind it will return Ok(None)
+        assert!(admin_t.is_none() == true);
```
Comment (NavinShrinivas) [inline server/lib/src/idm/ldap.rs:666]: this test is failing

File: server/lib/src/server/mod.rs [hunk 14]
```diff
@@ -1500,8 +1513,33 @@ impl<'a> QueryServerWriteTransaction<'a> {
     pub(crate) fn reload_domain_info(&mut self) -> Result<(), OperationError> {
         let domain_name = self.get_db_domain_name()?;
         let display_name = self.get_db_domain_display_name()?;
+        let domain_ldap_allow_unix_pw_bind = match self.get_domain_ldap_allow_unix_pw_bind() {
```
Comment (Firstyear) [inline server/lib/src/server/mod.rs:1516]: Here you're trying represent "the possibility of abscence" but you're encoding that into the Result type. Result shows "the possibility of error".

If you want to still do it this way, you want to change the type of get_domain_ldap_allow_unix_pw_bind to `Result<Option<bool>, OperationError>`. This way you show that there is a possibility of error, and if there was no error, that the contained value has a possibility of absecence. Then you con do "unwrap_or(true)" which allows you to set a default value here.

File: server/lib/src/server/mod.rs [hunk 15]
```diff
@@ -1500,8 +1513,33 @@ impl<'a> QueryServerWriteTransaction<'a> {
...
+            Ok(v) => {
+                v
+            },
+            _ => {
+                admin_warn!("Defaulting ldap_allow_unix_pw_bind to true");
+
+                let modl = ModifyList::new_purge_and_set(
```
Comment (Firstyear) [inline server/lib/src/server/mod.rs:1523]: Don't do this internal_modify here - you are *reading* the state of the boolean, not updating it. The reason for this is to consider "on upgrade defaults".

If you *write* to the database, we can't tell if the write was from us or the user, so we have to remain stuck on the default.

But if we allow the bool to be absent from the entry, and we just "unwrap_or(true)", then we can later change the "absent" state from true -> false, and every user gets the default changed. Had you written that to the db we can't do that!

File: server/lib/src/server/mod.rs [hunk 16]
```diff
@@ -1500,8 +1513,33 @@ impl<'a> QueryServerWriteTransaction<'a> {
...
+                    Err(e) => {
+                       return Err(e);
+                    }
+                }
+                true
+            }
+        };
         let domain_uuid = self.be_txn.get_db_d_uuid();
         let mut_d_info = self.d_info.get_mut();
+        if mut_d_info.d_ldap_allow_unix_pw_bind != domain_ldap_allow_unix_pw_bind {
```
Comment (Firstyear) [inline server/lib/src/server/mod.rs:1540]: You don't need this check, just always write it to the structure.

File: server/lib/src/idm/ldap.rs [hunk 17]
```diff
@@ -642,16 +645,35 @@ mod tests {
...
+
+        // ALLOWING UNIX_PW FOR BIND :
+        let mut idms_prox_write = idms.proxy_write(duration_from_epoch_now()).await;
```
Comment (Firstyear) [inline server/lib/src/idm/ldap.rs:669]: Shouldn't you be setting this to "false" to show what happens if you make it false?

File: server/lib/src/idm/server.rs (hunk 12 above)
Comment (Firstyear) [reply server/lib/src/idm/server.rs:1364]: Just set the logic to "what it should be", even if that causes test failures for now.

**Review by Firstyear — CHANGES_REQUESTED** [review]

File: server/lib/src/idm/ldap.rs (hunk 17 above)
Comment (Firstyear) [reply server/lib/src/idm/ldap.rs:669]: I think your logic here is fliipped btw. when the value is *absent* unix bind should work.
when allow pw bind is true it should work
and when it's false, it should *not* work.

So because you're never setting false, it isn't working I think. I think you need to more clearly seperate out the three cases in the test to match the states that I mentioned (about possibility of abscence, then the two possible bool states).

File: server/lib/src/server/mod.rs (hunk 14 above)
Comment (NavinShrinivas) [reply server/lib/src/server/mod.rs:1516]: But if I default the `get` method to return false on absence of an error then i wouldnt have to default anything in this relaod function right?

File: server/lib/src/server/mod.rs (hunk 15 above)
Comment (NavinShrinivas) [reply server/lib/src/server/mod.rs:1523]: Ahhh, had my suspicions that the writing defaults to database was causing some messups...this makes sense!

File: server/lib/src/server/mod.rs (hunk 14 above)
Comment (Firstyear) [reply server/lib/src/server/mod.rs:1516]: Correct. I was more explaining there are two ways to approach it :)

Comment (Firstyear) [reply server/lib/src/server/mod.rs:1516]: (Also you want to return true by default as allow unix pw bind is the default for now).

**Comment by NavinShrinivas** [comment]
@Firstyear Should Anon binds happen even if unix binding is not allowed? If I understand right anon binding can be used in kanidm to read the details off of the directory, right?

**Comment by Firstyear** [comment]
> @Firstyear Should Anon binds happen even if unix binding is not allowed? If I understand right anon binding can be used in kanidm to read the details off of the directory, right?

This is an excellent question.

I think the answer is "yes anonymous binds should be allowed" because anonymous binds are a special case in ldap. The goal of this change is to isolate and reduce the use of the unix pw. Since that's the goal, anonymous binds and their existance don't block that.

Anonymous binds can of course be stopped by an alternate control (locking the anonymous account).

**Comment by NavinShrinivas** [comment]
> answer is "yes anonymous binds

Ah great! Test driven developer caught this case for me :). Anyways, this will require me to write a logic checking if the UUID is that of `Anon` user that skip the flag check, I feel this is addition `not clean logic`. Any alternatives?

**Comment by Firstyear** [comment]
We already have to special case the anonymous bind anyway in the simple bind path, so I think it won't require any extra code on your part - just put the check for "is unix bind allowed" in the branch for "when not anonymous".

**Comment by NavinShrinivas** [comment]
Ah got it! Also for tests, I've written the following atm : 

using default flag : 
- `admin` as dn (passes)
- `admin@example.com` as dn (passes) 

using flag as false : 
- anon bind | empty dn and pw (passes) 
- empty dn, non-empty pw (not authenticated, does not pass)
- normal unix bind with `admin` dn (doesn't pass) 

using flag as true : 
- the original set of tests (where all pass) 

I've not tested the default and false flag under every possible DN and pw possibility. Does this suffice?

File: server/lib/src/constants/entries.rs [hunk 18]
```diff
@@ -361,6 +363,7 @@ impl From<Attribute> for &'static str {
     fn from(val: Attribute) -> Self {
         match val {
             Attribute::Account => ATTR_ACCOUNT,
+            Attribute::DomainLdapAllowUnixPwBind => ATTR_DOMAIN_LDAP_ALLOW_UNIX_PW_BIND,
```
Comment (Firstyear) [inline server/lib/src/constants/entries.rs:366]: Should be sorted alphabetically in the list

File: server/lib/src/constants/uuids.rs [hunk 19]
```diff
@@ -237,6 +237,8 @@ pub const UUID_SCHEMA_ATTR_AUTH_SESSION_EXPIRY: Uuid =
     uuid!("00000000-0000-0000-0000-ffff00000141");
 pub const UUID_SCHEMA_ATTR_AUTH_PRIVILEGE_EXPIRY: Uuid =
     uuid!("00000000-0000-0000-0000-ffff00000142");
+pub const UUID_SCHEMA_ATTR_DOMAIN_LDAP_ALLOW_UNIX_PW_BIND: Uuid =
```
[the remark this answers is not in the document]
Comment (Firstyear) [reply server/lib/src/constants/uuids.rs:240]: Has this been rebase to master? This is a uuid conflict.

File: server/lib/src/server/migrations.rs [hunk 20]
```diff
@@ -594,6 +595,8 @@ impl<'a> QueryServerWriteTransaction<'a> {
         // First, check the system_info object. This stores some server information
         // and details. It's a pretty const thing. Also check anonymous, important to many
         // concepts.
+
+        // The below only check if the UUID of the classes exists (Should exists from migrations)
```
Comment (Firstyear) [inline server/lib/src/server/migrations.rs:599]: What does this comment mean?

File: server/lib/src/server/mod.rs [hunk 21]
```diff
@@ -1500,8 +1511,31 @@ impl<'a> QueryServerWriteTransaction<'a> {
     pub(crate) fn reload_domain_info(&mut self) -> Result<(), OperationError> {
         let domain_name = self.get_db_domain_name()?;
         let display_name = self.get_db_domain_display_name()?;
+        let domain_ldap_allow_unix_pw_bind = match self.get_domain_ldap_allow_unix_pw_bind() {
+            Ok(v) => v,
+            _ => {
+                admin_warn!("Defaulting ldap_allow_unix_pw_bind to true");
+                info!("Defaulting ldap_allow_unix_pw_bind to true");
```
Comment (Firstyear) [inline server/lib/src/server/mod.rs:1518]: I think you only need one log message here. For now make it debug.

File: server/lib/src/server/mod.rs [hunk 22]
```diff
@@ -1500,8 +1511,31 @@ impl<'a> QueryServerWriteTransaction<'a> {
...
+
+                // let modl = ModifyList::new_purge_and_set(
+                //     Attribute::DomainLdapAllowUnixPwBind,
+                //     Value::Bool(true),
```
Comment (Firstyear) [inline server/lib/src/server/mod.rs:1522]: Remove commented code here.

File: server/lib/src/server/mod.rs [hunk 23]
```diff
@@ -1630,7 +1665,7 @@ impl<'a> QueryServerWriteTransaction<'a> {
     #[instrument(level = "info", skip_all)]
     pub fn commit(mut self) -> Result<(), OperationError> {
         self.reload()?;
-
+        info!("here : commiting qs_write");
```
Comment (Firstyear) [inline server/lib/src/server/mod.rs:1668]: delete log message, it's already implied by the instrument line.

File: server/lib/src/server/mod.rs [hunk 24]
```diff
@@ -793,6 +794,15 @@ pub trait QueryServerTransaction<'a> {
             })
     }
 
+    fn get_domain_ldap_allow_unix_pw_bind(&mut self) -> Result<bool, OperationError> {
+        let res = self.internal_search_uuid(UUID_DOMAIN_INFO)?;
+        match res.get_ava_single(Attribute::DomainLdapAllowUnixPwBind) {
```
Comment (Firstyear) [inline server/lib/src/server/mod.rs:799]: Change to
```
self.internal_search_uuid(UUID_DOMAIN_INFO)
    .map(|entry| {
        entry.get_ava_single_bool(Attribute::DomainLdapAllowUnixPwBind).unwrap_or(true)
    })
```

File: server/lib/src/idm/ldap.rs [hunk 25]
```diff
@@ -643,15 +642,51 @@ mod tests {
         let pce = UnixPasswordChangeEvent::new_internal(UUID_ADMIN, TEST_PASSWORD);
 
         assert!(idms_prox_write.set_unix_account_password(&pce).is_ok());
-        assert!(idms_prox_write.commit().is_ok());
+        assert!(idms_prox_write.commit().is_ok()); // Committing all configs
 
-        let anon_t = ldaps.do_bind(idms, "", "").await.unwrap().unwrap();
+        // Test for optional use of POSIX password for bind
+        // Empty UNIX_PW bind, should allow :
```
Comment (Firstyear) [inline server/lib/src/idm/ldap.rs:648]: Comment here doesn't seem to match the test? Should this mean "allow_unix_pw_bind unset - > binds allowed"?

File: server/lib/src/idm/ldap.rs [hunk 26]
```diff
@@ -643,15 +642,51 @@ mod tests {
...
+        let admin_t = ldaps
+            .do_bind(idms, "admin", TEST_PASSWORD)
+            .await
+            .unwrap()
+            .unwrap();
+        assert!(admin_t.effective_session == LdapSession::UnixBind(UUID_ADMIN));
+        let admin_t = ldaps
```
Comment (Firstyear) [inline server/lib/src/idm/ldap.rs:655]: Why do this twice?

File: server/lib/src/idm/ldap.rs [hunk 27]
```diff
@@ -643,15 +642,51 @@ mod tests {
...
+            .do_bind(idms, "admin@example.com", TEST_PASSWORD)
+            .await
+            .unwrap()
+            .unwrap();
+        assert!(admin_t.effective_session == LdapSession::UnixBind(UUID_ADMIN));
+
+        // Setting UNIX_PW to false:
```
Comment (Firstyear) [inline server/lib/src/idm/ldap.rs:662]: As above, improve comment.

**Review by Firstyear — CHANGES_REQUESTED** [review]

File: server/lib/src/constants/entries.rs (hunk 18 above)
Comment (NavinShrinivas) [reply server/lib/src/constants/entries.rs:366]: Done

File: server/lib/src/server/migrations.rs (hunk 20 above)
Comment (NavinShrinivas) [reply server/lib/src/server/migrations.rs:599]: It was a personal note to keep track of code. Removing it

File: server/lib/src/server/mod.rs (hunk 23 above)
Comment (NavinShrinivas) [reply server/lib/src/server/mod.rs:1668]: Sorry, all the info!() was to debug stuff in testing....removing them

File: server/lib/src/idm/ldap.rs (hunk 25 above)
Comment (NavinShrinivas) [reply server/lib/src/idm/ldap.rs:648]: Yes, teh variables names are wrong, making them right

File: server/lib/src/server/mod.rs (hunk 24 above)
Comment (NavinShrinivas) [reply server/lib/src/server/mod.rs:799]: why tho? can we have many entries? I don't understand why there might be differentiating entries where each define changes to a subset of variables? Does each schema become one entry value?

Comment (NavinShrinivas) [reply server/lib/src/server/mod.rs:799]: Also what if internak_search_uuid gives out an error?

Comment (Firstyear) [reply server/lib/src/server/mod.rs:799]: That's handled. self.internal_search_uuid() -> returns a Result<Entry, Error>
.map -> for the value of the Ok(entry), get the attribute assuming it's a single-boolean. 
   unwrap_or tream means "if no attribute matechs, or it's the wrong type, return true"

So this handles those cases.

File: server/lib/src/idm/ldap.rs (hunk 26 above)
Comment (NavinShrinivas) [reply server/lib/src/idm/ldap.rs:655]: they are on different DNs, just wanted to include those in test suites as well....can remove it

File: server/lib/src/server/mod.rs [hunk 28]
```diff
@@ -814,6 +815,16 @@ pub trait QueryServerTransaction<'a> {
             })
     }
 
+    fn get_domain_ldap_allow_unix_pw_bind(&mut self) -> Result<bool, OperationError> {
+        self.internal_search_uuid(UUID_DOMAIN_INFO).map(|entry| {
+            entry
+                .get_ava_single(Attribute::DomainLdapAllowUnixPwBind)
+                .unwrap_or(Value::new_bool(true))
```
Comment (Firstyear) [inline server/lib/src/server/mod.rs:822]: use `get_ava_single_bool()` which already does this for you.

File: server/lib/src/idm/server.rs [hunk 29]
```diff
@@ -1379,6 +1379,10 @@ impl<'a> IdmServerAuthTransaction<'a> {
                 effective_session: LdapSession::UnixBind(UUID_ANONYMOUS),
             }))
         } else {
+            if self.qs_read.d_info.d_ldap_allow_unix_pw_bind == false {
```
Comment (Firstyear) [inline server/lib/src/idm/server.rs:1382]: clippy will complain, this should be `if !self.qs_read.d_info.d_ldap_allow_unix_pw_bind { `

File: server/lib/src/idm/server.rs [hunk 30]
```diff
@@ -1379,6 +1379,10 @@ impl<'a> IdmServerAuthTransaction<'a> {
...
+                security_info!("Bind not allowed through Unix passwords.");
+                return Ok(None);
```
Comment (Firstyear) [inline server/lib/src/idm/server.rs:1384]: Probably should return an invalid credential error instead of Ok(None) here.

File: server/lib/src/idm/ldap.rs (hunk 13 above)
Comment (Firstyear) [reply server/lib/src/idm/ldap.rs:680]: Why is it failing? What's being returned here? What's happening in the logs?

**Review by Firstyear — CHANGES_REQUESTED** [review]

**Comment by NavinShrinivas** [comment]
Don't know why, ghub is not letting me reply on that last comment. None of the tests are failing....I've just written comments telling that the bind wont happen as the flag is set to false. On non binding replies is Ok(None) accepted? (Considering your second last comment, this is not the case and infact invalid credentials error  is expected)

**Comment by NavinShrinivas** [comment]
> Don't know why, ghub is not letting me reply on that last comment. None of the tests are failing....I've just written comments telling that the bind wont happen as the flag is set to false. On non binding replies is Ok(None) accepted? (Considering your second last comment, this is not the case and infact invalid credentials error is expected)

The reason I use Ok(None) is because all other cases in `auth_ldap()` returns back Ok(None) (say for example, not having posix password configured)

**Comment by NavinShrinivas** [comment]
And furthermore, do_op converts Ok(None) from do_bind into InvalidCredentails and most code invokes binding through do_op. If we were to change what do_bind sends out, should we change do_op as well?

**Comment by NavinShrinivas** [comment]
image
And do_op seems to be expection Ok(None) where it generates the error?

**Comment by Firstyear** [comment]
My mistake, you're right :) Ok(None) is correct here

File: server/lib/src/idm/ldap.rs [hunk 31]
```diff
@@ -643,15 +642,54 @@ mod tests {
...
+        assert!(admin_t.effective_session == LdapSession::UnixBind(UUID_ADMIN));
+        let admin_t = ldaps
+            .do_bind(idms, "admin@example.com", TEST_PASSWORD)
+            .await
+            .unwrap()
+            .unwrap();
+        assert!(admin_t.effective_session == LdapSession::UnixBind(UUID_ADMIN));
 
+        // Setting UNIX_PW_BIND flag to false:
+        // Hence all of the below tests will fail (asserts are still satisfied)
```
Comment (Firstyear) [inline server/lib/src/idm/ldap.rs:663]: Maybe reword to "authentication will fail" rather than the test failing. The test is succeeding at showing auth failing!

**Comment by NavinShrinivas** [comment]
Hey, do we have to fix these failing builds?

Also, I'd like to add docs regarding this option and its behaviour. Is there any guidelines for writing these docs?

**Comment by Firstyear** [comment]
If the failures are because of something in the PR, then yeah, it needs to be fixed.

As for the docs, leave that to me I think. :)

**Comment by Firstyear** [comment]
I went to check out and edit your branch but it looks completely cooked - I have no idea how you were working on this tbh. Going to fix it in a new pr.

**Comment by NavinShrinivas** [comment]
> I went to check out and edit your branch but it looks completely cooked - I have no idea how you were working on this tbh. Going to fix it in a new pr. 
> 
> EDIT: It's because you are working on your forks master branch @NavinShrinivas  - that's mucking up a whole bunch of things if anyone else tries to clone your fork and then review your work. So It's probably a good future thing to "never do this". 

Yes I felt this pr was really messy too, do you want me to fix this? I'd like to open a proper pr that's cleaner

**Comment by Firstyear** [comment]
> > I went to check out and edit your branch but it looks completely cooked - I have no idea how you were working on this tbh. Going to fix it in a new pr.
> > EDIT: It's because you are working on your forks master branch @NavinShrinivas  - that's mucking up a whole bunch of things if anyone else tries to clone your fork and then review your work. So It's probably a good future thing to "never do this".
> 
> Yes I felt this pr was really messy too, do you want me to fix this? I'd like to open a proper pr that's cleaner

It's going to be super hard to fix. You can't rebase-squash to get a single commit here because you did merge commits. I think you have to do a shit ton of rebasing to rearrance history and make it work. I'm thinking the easiest way will be a new branch, completely clean from master, and copy-paste your changes from the diff viewer here on github.

Trying to resolve this in git is going to be simply too painful I think.

**Comment by NavinShrinivas** [comment]
> diff

What about me cloning you upstream, cherry picking my commits from my fork and pushing that as a new branch onto my fork and I'll make a pr from there

**Comment by Firstyear** [comment]
That would work, but then you have to cherry pick 11 commits, doing a manual conflict resolve at every step.

**Comment by Firstyear** [comment]
This was merged in a separate PR.
